Implement "User" data type

zackkrida opened the pull request:
This PR adds a user data type to Mathesar. This type can be used to track the user who last edited a column via the Mathesar UI. This editing is not tracked outside of the Mathesar UI, so is not currently meant to be used as a robust form of auditing. 

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

zackkrida committed:
Add user type column support to backend

- Add user_type metadata field to ColumnMetadata model
- Add user type column handling in RPC endpoints (columns, records, forms)
- Add user display utilities for formatting user data
- Add migration for user_type column metadata
- Make user_display_field nullable to match pattern of other type-specific fields

zackkrida committed:
Add user type column support to frontend

- Add UserCell, UserInput, and UserFilterInput components
- Add user type to abstract type system and type definitions
- Add user type column display options (username, display name, email)
- Add user type filtering support
- Add user type default value options (auto-set to editor, default user)
- Prevent user type columns from being added to forms
- Add user utilities for fetching and displaying users

zackkrida committed:
Add user type documentation

- Add comprehensive user type guide with examples
- Add user type images for documentation
- Update data types documentation to include user type

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida commented:
@mathemancer please review the backend implementation in [`6c34b8c` (#4997)](https://github.com/mathesar-foundation/mathesar/pull/4997/commits/6c34b8ceff4475d759847ab820e97ff61bbaacc8). 

Regarding the frontend, I need to compare this to some overlapping work in cross-table editing to validate the approach, and make this a bit easier to review for the frontend team before getting frontend review.

mathemancer commented on mathesar/models/base.py:
```diff
@@ -239,6 +239,13 @@ class ColumnMetaData(BaseModel):
+        max_length=50,
+        null=True
+    )
+    user_last_edited_by = models.BooleanField(default=False, null=True)
```
I find this name really confusing. What about `track_editing_user`?

mathemancer commented on mathesar/rpc/columns/metadata.py:
```diff
@@ -50,6 +53,10 @@ class ColumnMetaDataRecord(TypedDict):
     duration_min: Optional[str]
     duration_max: Optional[str]
     display_width: Optional[int]
+    file_backend: Optional[str]
```
Oops, nice catch

mathemancer commented on mathesar/rpc/columns/metadata.py:
```diff
@@ -50,6 +53,10 @@ class ColumnMetaDataRecord(TypedDict):
     display_width: Optional[int]
+    file_backend: Optional[str]
+    user_type: Optional[bool]
+    user_display_field: Optional[Literal["full_name", "email", "username"]]
```
If `user_display_field` is optional, I guess I don't see why `user_type` is needed. If this is non-null, then display the column as specified, otherwise, just display as an integer.

If you really really want to keep `user_type`, I _don't_ think that should be optional since it's always either true or false. Just use `false` for the default.

zackkrida replied:
Got rid of user_type 👍

mathemancer commented on mathesar/rpc/columns/metadata.py:
```diff
@@ -72,6 +79,8 @@ def from_model(cls, model):
             file_backend=model.file_backend,
+            user_type=model.user_type,
+            user_display_field=model.user_display_field or "username",
         )
```
I think `user_last_edited_by` is missing here.

zackkrida replied:
Covered by `track_editing_user` instead.

mathemancer commented on mathesar/rpc/forms.py:
```diff
@@ -414,4 +414,7 @@ def submit(*, form_token: str, values: dict, **kwargs) -> None:
-    return submit_form(form_token, values)
+    from modernrpc.core import REQUEST_KEY
+    request = kwargs.get(REQUEST_KEY)
+    user = request.user if request else None
```
- Why are you reimporting `REQUEST_KEY`?
- This isn't going to be called without a request, right? If not, there's way too much logic here.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -233,16 +342,16 @@ def from_dict(cls, d):
+    filter: Filter = None,
+    grouping: Grouping = None,
+    return_record_summaries: bool = False,
+    **kwargs,
```
It seems like your editor really hates the double-indent function args style, but I prefer it. This comment is applicable to many of your PRs (maybe all)

zackkrida replied:
Yeah I'll disable automatic formatting in vscode for Mathesar. Preferences aside it clutters the diffs too much anyway.

zackkrida replied:
Yeah I'll disable automatic formatting in vscode for Mathesar. Preferences aside it clutters the diffs too much anyway.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -365,26 +487,33 @@ def add(
     """
     user = kwargs.get(REQUEST_KEY).user
+    # Automatically set user_last_edited_by columns to current user ID
+    _set_last_edited_by_columns(record_def, table_oid, database_id, user.id)
```
This should be in a separate module, and should not modify records in place.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -273,27 +382,34 @@ def list_(
+        or None
+    )
+
+    _add_user_display_values_to_record_info(record_info, table_oid, database_id)
```
This should be in a separate module, and should not modify in place.

mathemancer reviewed (changes requested):
Partial review so you have stuff to think about:

- Please avoid hiding the mutation of python objects in function calls, as you do in `_add_user_display_values_to_record_info` et al. I'd really prefer the modifications be inline where they're clearly visible; see the addition of `download_links` for example.
- You need to refactor this so that a given RPC call never generates multiple requests for the same Django model object. For example, `records.add` calls `get_columns_meta_data` in both `_set_last_edited_by_columns` and `_add_user_display_values_to_record_info`. The problem is also present in `records.list` since both `get_download_link_columns` and `_add_user_display_values_to_record_info` call the `get_columns_meta_data` function. We either need to cache that metadata getting function for the duration of a request, or (easier) just call it in the main `rpc` function body then use it where needed.

More to come later; late for the team event!

zackkrida committed:
Address review feedback from Brent

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida committed:
Resolve linting and formatting issues

zackkrida commented:
@mathemancer I addressed your initial feedback some time ago but never requested re-review. This should be good for you to look at once more. 

@pavish, likewise for the frontend. I haven't revisited that work in some time so I imagine there are some aspects that aren't ideal. I'll be available to make changes as you see fit and happy to do any major refactoring. 

@mathemancer one thing I wanted to mention. @kgodey and I discussed that we'd like to hold off on making the user edit tracking and auditing more robust or enforced at the DB level. I could get into reasons why, but the main one is "scope" and "we can do it later if requested". Feel free to reach out to either of us for more details though if you'd like to push back on that.

zackkrida committed:
Run Prettier on frontend files

zackkrida committed:
Fix type errors

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida commented:
@pavish and @mathemancer I finally had some time to give this a deeper review, it's not ready for your reviews yet. There's an obvious issue with migrations and some frontend behavior I'd like to fix ASAP. I'll undraft this when it's ready for review once more.

zackkrida committed:
Fixed migration conflict (duplicate 0010_), updated tests for new user-type metadata fields, fixed UserCell to call setValue() instead of dispatching an unlistened update event, replaced broken get() reactivity with proper Svelte store subscriptions, extracted duplicated user-record utilities into shared module, and other minor fixes.

zackkrida committed:
Fix migration order, fix user retrieval on anonymous routes, and fix  submit_form track-editor behavior

zackkrida committed:
Fix lint and type errors

zackkrida committed:
Merge branch 'develop' into user-type

mathemancer commented on mathesar/rpc/columns/metadata.py:
```diff
@@ -96,6 +104,9 @@ class ColumnMetaDataBlob(TypedDict):
         file_backend: The name of a backend for storing file attachments.
+        user_display_field: Which user field to display for user columns (full_name, email, or username).
+            If non-null, the column is treated as a user column.
+        track_editing_user: Whether this column automatically tracks the last user who edited the record.
```
This is a bit fishy: What happens if the columns _isn't_ a `user_display_field`, or the underlying type is inappropriate? What happens to `track_editing_user`? Should that logic run, or is it predicated on both being true?

mathemancer commented on mathesar/migrations/0011_columnmetadata_user_type.py:
```diff
@@ -0,0 +1,31 @@
+        migrations.AddField(
+            model_name="columnmetadata",
+            name="track_editing_user",
+            field=models.BooleanField(default=False),
```
Seems like it should not be possible (logically) to set `track_editing_user` to true unless `user_display_field` is true. You can accomplish this with

```sql
-- track_editing_user IS true => user_display_field IS NOT NULL
CHECK (NOT track_editing_user OR user_display_field IS NOT NULL)
```

This would make the data model a bit more robust.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -316,6 +333,7 @@ def list_(
         The requested records, along with some metadata.
     """
     user = kwargs.get(REQUEST_KEY).user
+    columns_meta_data = list(get_columns_meta_data(table_oid, database_id))
```
Good idea combining these DB hits.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -327,15 +345,28 @@ def list_(
     ) or None
 
+    # Add user display values for user columns
+    user_summaries = get_user_linked_record_summaries(
+        columns_meta_data, record_info.get("results", [])
+    )
```
The inconsistency between this and the `get_download_links` call is a bit offputting: In one case we only pass along the columns for which we need the download links, but in the other we pass along all metadata and hope (from this perspective) that the function called is smart enough to only generate summaries for appropriate columns. Given my concerns about this function knowing too much about the definition of `download_link_columns`, I think we should do the latter, so the `get_download_links` function should accept the entire metadata blob.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -327,15 +345,28 @@ def list_(
-    download_link_columns = get_download_link_columns(table_oid, database_id)
+
+    # Add download links for file columns
+    download_link_columns = [c.attnum for c in columns_meta_data if c.file_backend]
```
I don't like this function knowing which attribute constitutes a "download_link_column".

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -385,12 +416,24 @@ def get(
+    user_summaries = get_user_linked_record_summaries(
+        columns_meta_data, record_info.get("results", [])
+    )
+    if user_summaries:
```
Same problem here.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -327,15 +345,28 @@ def list_(
+    user_summaries = get_user_linked_record_summaries(
+        columns_meta_data, record_info.get("results", [])
+    )
+    if user_summaries:
```
```suggestion
    if user_summaries is not None:
```
protects against falsy but non-iterable outputs of `get_user_linked_record_summaries`.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -327,15 +345,28 @@ def list_(
+    if user_summaries:
+        if record_info.get("linked_record_summaries") is None:
+            record_info["linked_record_summaries"] = {}
+        record_info["linked_record_summaries"].update(user_summaries)
```
```suggestion
        record_info.setdefault("linked_record_summaries", {}).update(user_summaries)

```
Avoid nested ifs for maximum pythonicity

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -385,12 +416,24 @@ def get(
+    if user_summaries:
+        if record_info.get("linked_record_summaries") is None:
+            record_info["linked_record_summaries"] = {}
+        record_info["linked_record_summaries"].update(user_summaries)
```
as above.

mathemancer commented on mathesar/rpc/records.py:
```diff
@@ -424,14 +467,29 @@ def add(
+    if user_summaries:
+        if record_info.get("linked_record_summaries") is None:
+            record_info["linked_record_summaries"] = {}
+        record_info["linked_record_summaries"].update(user_summaries)
```
As above.

mathemancer commented on mathesar/utils/user_display.py:
```diff
@@ -0,0 +1,99 @@
+    for column_attnum, display_field in user_columns:
+        user_ids = set()
+        for record in results:
+            user_id = record.get(str(column_attnum)) or record.get(column_attnum)
```
Is this check necessary? If it is, there's a bug in `apply_track_editing_user`: If the column reference is an integer, you'll add a new column in that record with a string key (instead of integer).

mathemancer commented on mathesar/utils/user_display.py:
```diff
@@ -0,0 +1,99 @@
+    result = dict(record_def)
+    for col in columns_meta_data:
+        if col.user_display_field and col.track_editing_user:
+            result[str(col.attnum)] = user_id
```
Potential bug if it's possible for the result keys to be integers.

mathemancer commented on mathesar/utils/forms.py:
```diff
@@ -220,6 +232,66 @@ def submit_form(form_token, values):
     ]
+
+    # Set track_editing_user columns for authenticated users
+    if user and user.is_authenticated and hasattr(user, 'id'):
```
All of this logic should be extracted to other functions.

mathemancer commented on mathesar/utils/forms.py:
```diff
@@ -204,9 +205,20 @@ def iterate_form_fields(fields, parent_field=None, depth=0, fields_to_pick=[]):
+    """
     form_model = Form.objects.get(token=form_token)
     assert form_model.publish_public, 'This form does not accept submissions'
+    values = dict(values)
```
What is this protecting against? Did you encounter errors without it?

mathemancer commented on mathesar/utils/forms.py:
```diff
@@ -220,6 +232,66 @@ def submit_form(form_token, values):
+                    "table_oid": field_info['table_oid'],
+                    "parent_key": field_info['parent_key'],
+                    "depth": field_info['depth'],
+                })
```
What's the point of both of these objects? Seems like only one or the other should exist. Why not just the set of tuples? What is the `dict` bringing to the table?

mathemancer commented on mathesar/utils/forms.py:
```diff
@@ -220,6 +232,66 @@ def submit_form(form_token, values):
+                    "parent_key": parent_key,
+                    "column_attnum": column_attnum,
+                    "table_oid": table_oid,
+                    "depth": depth
```
This part makes me realize that `track_editing_user` should be a table property, not a column one. In particular, there should be a choice at the table level of which column we should auto-insert user info into, and the column metadata should purely refer to display style. This has benefits for the user, but also makes most of this logic extraneous.

zackkrida replied:
I really appreciate this suggestion, it dramatically improved the PR's code and user experience.

mathemancer reviewed (dismissed):
High-level things I noticed:

- When setting the display from Username to, e.g., email, the UI doesn't update to show the new display. This is contrast to other display styles, and breaks user expectations driven by those other interactions without warning.
- Handles inappropriate values in user-display integer column (e.g., entered manually via psql) gracefully by degrading to integer display. Consider some kind of "broken link" display similar to how we handle file data type columns when the thumbnail is unfetchable. 
- I probably already pestered you about this but please avoid massive formatting overhauls intermingled with logic changes. It's a pain to review. I'd be open to choosing a formatter and running it on the whole python codebase so we can avoid the problem moving forward.
- It occurred to me over the course of reviewing this that the user flow is upside down, and discoverability is reduced by hiding "track editing users" deep in column metadata. Better to have that be table metadata, and have a flow for creating a column to store the info if needed. 

That last point makes it obvious (to me) that this feature was conceived of "feature-first" rather than "use-case-first". If the desired use-case is "track editors of table", then that should be a table setting, and where the editor ID is stored is sort of an implementation detail. The current setup asks a user to stumble across an implementation detail, choose to use that (the User column) without any indication that it's for tracking editors, then discover that they can leverage the type they chose for editor tracking and auditing.

zackkrida committed:
Address PR review feedback for user type feature

Move user tracking from column-level to table-level: replace
track_editing_user column flag with user_tracking_attnum on TableMetaData.
Add table inspector UI section "Edit Tracking" to configure which
User-type column auto-records the editing user on each row add/patch.

- Backend: add user_tracking_attnum to TableMetaData model and RPC layer,
  remove track_editing_user from ColumnMetaData, update apply_track_editing_user
  to use table-level setting, fix get_download_links to accept full metadata
  blob, fix linked_record_summaries null handling in records.py
- Frontend: remove track_editing_user from column display options,
  add isUserTrackingColumn to ProcessedColumn, add Edit Tracking inspector
  section, fix setDisplayOptions to dispatch columnPatched so cells
  refresh immediately when user_display_field changes
- Tests: update test_c_metadata and test_user_display for new API

zackkrida committed:
Fix Prettier formatting and remove unnecessary type assertion

zackkrida committed:
Fix failing tests for user tracking and download links changes

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida committed:
Fix unused variable warning in test lambdas

zackkrida committed:
Address remaining PR review feedback from Brent

  - records.py: flatten nested user_summaries merge into a single
    dict-spread expression across all 5 RPC methods
  - forms.py: drop redundant hasattr(user, 'id') check
  - UserCell.svelte: render a broken-link fallback (warning icon +
    strikethrough ID) when a user_display column references a
    non-existent user, mirroring the File cell's unfetchable state

Merge remote-tracking branch 'origin/develop' into user-type

zackkrida committed:
Fix formatting of UserCell

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida committed:
Merge branch 'develop' into user-type

zackkrida committed:
finish revert adding database_id to roles.configured.list

zackkrida committed:
Fix code relating to roles.configured.list recent changes on develop

zackkrida committed:
Update user data type guide to reflect new table-level tracking

zackkrida committed:
Switch to "Record" instead of "Row" in the user type guide
